# Third deduplication

A final standardizing pass on **workmeta**, followed by a temporal filter to produce **contemporaryworkmeta**.


In [1]:
import pandas as pd

In [2]:
work = pd.read_csv('newworkmeta.tsv', sep = '\t', low_memory = False)

### Standardize names

In our second deduplication notebook, we wrote to file pairs of names that we discovered in the process of deduplicating works. The texts and titles of certain works were similar enough to convince us that these author names were also equivalent.

However, we don't know which name is better. To decide, we need some rules, which I will proceed to provide.

In [3]:
authsets = pd.read_csv('authorsets.tsv', sep = '\t', header = None, names = ['name1', 'name2'])
authsets.head()

name1                                      name2
0           Robinson, Mary Stephens                           Robinson, Mary S
1                Howes, Edith Annie                               Howes, Edith
2  Williams, Elizabeth Whitney, Mrs                Williams, Elizabeth Whitney
3   Somerville, E. Œ. (Edith Œnone)  Somerville, E. &#xbf;. (Edith &#xbf;none)
4                      Pearce, Donn                             Pearce, Donald

In [4]:
def name2prefer(row):
    ''' A bunch of rules we can use to make up our mind.
    Basically, diacritical marks are preferred to garbage
    equivalents; otherwise longer forms are usually preferred.
    '''
    option1 = row.name1
    option2 = row.name2
    if '??' in option1 or '#xbf;' in option1:
        return option2
    elif '??' in option2 or '#xbf;' in option2:
        return option1
    elif 'from old catalog' in option1:
        return option2
    elif 'from old catalog' in option2:
        return option1
    elif '?_' in option1:
        return option2
    elif '?_' in option2:
        return option1
    elif ' ̌' in option1:
        return option2
    elif ' ̌' in option2:
        return option1
    elif '1' in option1:
        return option2
    elif '1' in option2:
        return option1
    elif '  ' in option1:
        return option2
    elif '  ' in option2:
        return option1
    elif 'è' in option1 or 'é' in option1:
        return option1
    elif 'è' in option2 or 'é' in option2:
        return option2
    elif len(option1) > len(option2):
        return option1
    else:
        return option2

authsets = authsets.assign(normative = authsets.apply(name2prefer, axis = 1))
authsets.head()

name1  \
0           Robinson, Mary Stephens   
1                Howes, Edith Annie   
2  Williams, Elizabeth Whitney, Mrs   
3   Somerville, E. Œ. (Edith Œnone)   
4                      Pearce, Donn   

                                       name2                         normative  
0                           Robinson, Mary S           Robinson, Mary Stephens  
1                               Howes, Edith                Howes, Edith Annie  
2                Williams, Elizabeth Whitney  Williams, Elizabeth Whitney, Mrs  
3  Somerville, E. &#xbf;. (Edith &#xbf;none)   Somerville, E. Œ. (Edith Œnone)  
4                             Pearce, Donald                    Pearce, Donald

#### now actually make the change


In [5]:
firstset = set(authsets.name1)
secondset = set(authsets.name2)
ctr = 0

def applynorm(name):
    global firstset, secondset, authsets, ctr
    
    if name in firstset:
        newname = str(authsets.loc[authsets.name1 == name, 'normative'].values[0])
        ctr += 1
        return newname
    elif name in secondset:
        newname = str(authsets.loc[authsets.name2 == name, 'normative'].values[0])
        ctr += 1
        return newname
    else:
        return name
    
work = work.assign(author = work.author.map(applynorm))  
print(ctr)

2899


That change was made in 2899 rows.

### Correct authordates

An error in an earlier notebook has caused the author "blank" or "no author name provided" to inherit authordates. These are not real.

In [7]:
ctr = 0

def be_agnostic(row):
    global ctr
    
    if pd.isnull(row.author) or len(row.author) < 2:
        ctr += 1
        return ''
    else:
        return row.authordate

work = work.assign(authordate = work.apply(be_agnostic, axis = 1))
print('Change made in ' + str (ctr) + ' rows.')

Change made in 8553 rows.


### Update last date of composition

We have a column **lastcomp** that's supposed to contain an inference about the last possible date of composition, based on certain values of **datetype** and **enddate,** plus authors' death dates.

But the dates of death have been updated since **lastcomp** was created. We may be able to improve that column a little before using it in the next stage of dedup.

In [8]:
ctr = 0

def update(row):
    global ctr 
    
    lastcomp = row.latestcomp
    if pd.isnull(lastcomp):
        lastcomp = float('nan')
    else:
        lastcomp = int(lastcomp)
        
    authordates = str(row.authordate).strip('[.]() ')
    
    if pd.isnull(authordates):
        if pd.isnull(lastcomp) or lastcomp > 2010:
            return float('nan')
        else:
            return lastcomp
        
    elif authordates.startswith('d.') or (len(authordates) > 7 and '-' in authordates):
        try:
            death = int(authordates[-4 : ])
        except:
            death = 0
    else:
        death = 0
    
    if (death > 1500 and death < 2010) and death < lastcomp:
        ctr += 1
        return death
    else:
        return lastcomp

work = work.assign(updated = work.apply(update, axis = 1))
print('Change made in ' + str (ctr) + ' rows.')

Change made in 1927 rows.


In [9]:
work['latestcomp'] = work['updated']
work.drop(labels = ['updated'], axis = 1, inplace = True)

In [10]:
def whethertokeep(row):
    infdate = int(row.inferreddate)
    lastcomp = row.latestcomp
    if pd.isnull(lastcomp):
        return True
    else:
        lastcomp = int(lastcomp)
    
    if infdate > (lastcomp + 25):
        return False
    else:
        return True

work = work.assign(earlyedition = work.apply(whethertokeep, axis = 1))

In [11]:
work.loc[work.author.str.startswith('Gaskell, E', na = False), : ]

docid                    oldauthor  \
249            nyp.33433074857115  Gaskell, Elizabeth Cleghorn   
8947    uiuo.ark+=13960=t9z03hc07  Gaskell, Elizabeth Cleghorn   
10281          njp.32101019691409  Gaskell, Elizabeth Cleghorn   
10426          nyp.33433074857149  Gaskell, Elizabeth Cleghorn   
10427          nyp.33433074857131  Gaskell, Elizabeth Cleghorn   
11082          nyp.33433074857362  Gaskell, Elizabeth Cleghorn   
12045          nyp.33433082340302  Gaskell, Elizabeth Cleghorn   
12064          nyp.33433074857156  Gaskell, Elizabeth Cleghorn   
12236          inu.30000007412517  Gaskell, Elizabeth Cleghorn   
12364                uc1.b3322505  Gaskell, Elizabeth Cleghorn   
12365                uc1.b3322506  Gaskell, Elizabeth Cleghorn   
12808          nyp.33433074857503  Gaskell, Elizabeth Cleghorn   
13408                uc1.b4105357  Gaskell, Elizabeth Cleghorn   
13504          nyp.33433074857354  Gaskell, Elizabeth Cleghorn   
13598   uiuo.ark+=13960=t8jd57r2q  Gaskell, Elizabeth Cleghorn   
13599   uiuo.ark+=13960=t43r17n3n  Gaskell, Elizabeth Cleghorn   
13600   uiuo.ark+=13960=t6057xr4s  Gaskell, Elizabeth Cleghorn   
13718                uc1.b3548115  Gaskell, Elizabeth Cleghorn   
14108    uc2.ark+=13960=t4sj1k525  Gaskell, Elizabeth Cleghorn   
14747   dul1.ark+=13960=t6d22nv15  Gaskell, Elizabeth Cleghorn   
14748   dul1.ark+=13960=t5r792w66  Gaskell, Elizabeth Cleghorn   
23374          nyp.33433074857453  Gaskell, Elizabeth Cleghorn   
25113                uc1.b4104225  Gaskell, Elizabeth Cleghorn   
25931            pst.000000466448  Gaskell, Elizabeth Cleghorn   
26081    uc2.ark+=13960=t15m68w3h  Gaskell, Elizabeth Cleghorn   
27268    loc.ark+=13960=t0xp7wk60  Gaskell, Elizabeth Cleghorn   
30045          mdp.39015057079579  Gaskell, Elizabeth Cleghorn   
38368          mdp.39015033170104  Gaskell, Elizabeth Cleghorn   
42927                uc1.b3556243  Gaskell, Elizabeth Cleghorn   
42928                uc1.b3556245  Gaskell, Elizabeth Cleghorn   
...                           ...                          ...   
95616            pst.000029836529  Gaskell, Elizabeth Cleghorn   
96151          mdp.39015014540002  Gaskell, Elizabeth Cleghorn   
96340          mdp.49015002026533  Gaskell, Elizabeth Cleghorn   
96341          mdp.49015002026541  Gaskell, Elizabeth Cleghorn   
96342          mdp.49015002026558  Gaskell, Elizabeth Cleghorn   
96343          mdp.49015002026566  Gaskell, Elizabeth Cleghorn   
96344          mdp.49015002026574  Gaskell, Elizabeth Cleghorn   
96345          mdp.49015002026582  Gaskell, Elizabeth Cleghorn   
96438            pst.000028738046  Gaskell, Elizabeth Cleghorn   
96439            pst.000028737988  Gaskell, Elizabeth Cleghorn   
96440            pst.000028738770  Gaskell, Elizabeth Cleghorn   
96441            pst.000028738718  Gaskell, Elizabeth Cleghorn   
96442            pst.000028738657  Gaskell, Elizabeth Cleghorn   
96517          uc1.32106001936472  Gaskell, Elizabeth Cleghorn   
96643          uc1.31158008456823  Gaskell, Elizabeth Cleghorn   
96646          uc1.31158011219846  Gaskell, Elizabeth Cleghorn   
96671          uc1.31175003826487  Gaskell, Elizabeth Cleghorn   
101596         mdp.39015002237058  Gaskell, Elizabeth Cleghorn   
106282         mdp.39015008542196  Gaskell, Elizabeth Cleghorn   
108347         mdp.39015019751562  Gaskell, Elizabeth Cleghorn   
116633         mdp.39015054134237  Gaskell, Elizabeth Cleghorn   
118310         coo.31924077744807  Gaskell, Elizabeth Cleghorn   
130095         mdp.39015056903092  Gaskell, Elizabeth Cleghorn   
133660         mdp.39076002745318  Gaskell, Elizabeth Cleghorn   
133661         inu.30000096457001  Gaskell, Elizabeth Cleghorn   
133662         inu.30000110371311  Gaskell, Elizabeth Cleghorn   
133663         inu.30000096457951  Gaskell, Elizabeth Cleghorn   
133664         inu.30000096457944  Gaskell, Elizabeth Cleghorn   
133665         inu.30000110371345  Gaskell, Elizabeth Cleghorn   
13

In [12]:
sum(work.earlyedition)

129023

In [13]:
work.to_csv('../workmeta.tsv', sep = '\t', index = False)

### Also correct authordate error in manifestationmeta

Same error we corrected above.

In [15]:
manifest = pd.read_csv('../manifestationmeta.tsv', sep = '\t', low_memory = False)

In [16]:
ctr = 0
manifest = manifest.assign(authordate = manifest.apply(be_agnostic, axis = 1))
print('Change made in ' + str (ctr) + ' rows.')

Change made in 8553 rows.


In [17]:
manifest.to_csv('../manifestationmeta.tsv', sep = '\t', index = False)